test: add jenkins/set_xfail.py

hzhou opened the pull request:
## Pull Request Description
It is odd to have xfail.conf and set_xfail script maintained in
different repository. This patch adds the script that applies
xfail.conf.

The current script works with the current xfail.conf format, which uses
sed. However, the script actually parses the regex pattern and
applies them directly without depending on sed. Once stablized, we can
further simplify the xfail.conf format.

For example, the current syntax is like this:
```
* * * ch4:ofi * sed -i "s+\(^idup_comm_gen .*\)+\1 xfail=ticket3794+g" test/mpi/threads/comm/testlist
```
It can be simplified as
```
* * * ch4:ofi * /^idup_comm_gen/   xfail=ticket3794   test/mpi/threads/comm/testlist
```

Run the script after `configure` with:
```
python test/mpi/main/jenkins/set_xfail.py -j $JOB_NAME -c $compiler -o $jenkins_configure -q $label -m $netmod -f $SRC/test/mpi/maint/jenkins/xfail.conf
```


<!--
Insert description of the work in this merge request (above this comment),
particularly focused on _why_ the work is necessary, not _what_ you did.
-->

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->
[skip warnings]
## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
The script will print the exact rules that is applied per test, facilitating debugging by checking console log. For example:
```
set_xfail states: mpich-main-ch3-centos/compiler=gnu,jenkins_configure=default,label=centos64 gnu default ch3:tcp centos64
test/mpi/coll/testlist.cvar:
    * * * * *: ^allred3.* env=MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM=gentran_ring.* -> xfail=ticket0
    * * * * *: ^allred4.* env=MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM=gentran_ring.* -> xfail=ticket0
    * * * * *: ^allred6.* env=MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM=gentran_ring.* -> xfail=ticket0
    * * * * *: ^allred6.* env=MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM=gentran_ring.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_doubling.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_doubling.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_doubling.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_halving.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_halving.* -> xfail=ticket0
    * * * * *: ^allgatherv4.* env=MPIR_CVAR_IALLGATHERV_INTRA_ALGORITHM=gentran_recexch_halving.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^red_scat_block2.* env=MPIR_CVAR_IREDUCE_SCATTER_BLOCK_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^redscat2.* env=MPIR_CVAR_IREDUCE_SCATTER_INTRA_ALGORITHM=gentran_recexch.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw2.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
    * * * * *: ^alltoallw1.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw2.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
    * * * * *: ^alltoallw_zeros.* env=MPIR_CVAR_IALLTOALLW_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
test/mpi/errors/coll/testlist:
    * * * * *: ^reducelength .* -> xfail=ticket3655
    * * * * *: ^ireducelength .* -> xfail=ticket3655
    * * * * *: ^allreducelength .* -> xfail=ticket3655
    * * * * *: ^iallreducelength .* -> xfail=ticket3655
    * * * * *: ^reduceop .* -> xfail=ticket3655
    * * * * *: ^ireduceop .* -> xfail=ticket3655
    * * * * *: ^gatherlength .* -> xfail=ticket3655
    * * * * *: ^igatherlength .* -> xfail=ticket3655
    * * * * *: ^scatterlength .* -> xfail=ticket3655
    * * * * *: ^iscatterlength .* -> xfail=ticket3655
    * * * * *: ^allgatherlength .* -> xfail=ticket3655
    * * * * *: ^iallgatherlength .* -> xfail=ticket3655
    * * * * *: ^alltoalllength .* -> xfail=ticket3655
test/mpi/errors/comm/testlist:
    * * * * *: ^intercomm_abort .* -> xfail=ticket0
test/mpi/threads/pt2pt/testlist:
    * * * ch3:tcp *: ^mt_iprobe_isendrecv .* -> xfail=issue4258
    * * * ch3:tcp *: ^mt_improbe_isendrecv .* -> xfail=issue4258
```

hzhou commented:
@wesbland @tarudoodi Do you recall what is the reason those gentran tests are xfailed? I wonder whether we can write the tests smarter so we don't have to clutter the xfail list.

tarudoodi commented:
@hzhou Some of these xfails with ticket#0 shouldn't fail now, because of the new `MPIR_CVAR_COLLECTIVE_FALLBACK=silent` CVAR we use for testing. This CVAR was added as a part of collective selection code. 

I have a concern about adding perl based script. During the hackathon last year, we had a discussion about the programming languages for new scripts being added to MPICH and I think it was in the favor of `bash` and `python`.

hzhou commented:
> @hzhou Some of these xfails with ticket#0 shouldn't fail now, because of the new `MPIR_CVAR_COLLECTIVE_FALLBACK=silent` CVAR we use for testing. This CVAR was added as a part of collective selection code.
>

@tarudoodi Sounds good. Could you verify which of those tests can be removed from xfails and maybe push a PR for it?

> I have a concern about adding perl based script. During the hackathon last year, we had a discussion about the programming languages for new scripts being added to MPICH and I think it was in the favor of `bash` and `python`.

Ok, I will convert it to python.

tarudoodi commented:
@hzhou  Will the scripts support both the xfail formats or at some point you will remove all the `sed` syntax?

tarudoodi commented:
About removing the xfails for ticket#0 type, if we do that we would be testing some algorithm twice and it would just add to the total test time without any benefits.

hzhou commented:
> @hzhou Will the scripts support both the xfail formats or at some point you will remove all the `sed` syntax?

It supports both xfail formats already. But stick to the `sed` syntax for now because we haven't migrated all the jenkins jobs yet. We'll do the format switch at some point later.

hzhou commented:
> About removing the xfails for ticket#0 type, if we do that we would be testing some algorithm twice and it would just add to the total test time without any benefits.

I see. So the tests won't fail but it will just be falling back to default algorithm? Should we enhance the generation script so not to generate those tests?

tarudoodi commented:
> I see. So the tests won't fail but it will just be falling back to default algorithm? Should we enhance the generation script so not to generate those tests?

So I double checked, in our testing we just removed the xfails, probably missed pushing that patch to `main`. Yeah, if we want to be very precise the test generation scripts would have to be modified.

hzhou commented:
> 
> 
> > I see. So the tests won't fail but it will just be falling back to default algorithm? Should we enhance the generation script so not to generate those tests?
> 
> So I double checked, in our testing we just removed the xfails, probably missed pushing that patch to `main`. Yeah, if we want to be very precise the test generation scripts would have to be modified.

Sounds good. I'll work on to modify the test generation scripts.

hzhou commented:
@tarudoodi Do you recall what is the reason these tests need fallback? I like to see if we can apply some rules to the test generation script.

tarudoodi commented:
I don't know for all of them, but in `gentran_ring` xfails, the tests have non-commutative ops.

tarudoodi commented:
For 
```
* * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
* * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
* * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
```
The algorithms don't support `in_place` sendbuf.  The `MPIR_Ialltoallv_impl`  function lists the exceptions to the algorithms. Same can be check in the other collectives/algos. I think the tests to be excluded are already listed here.

hzhou commented:
> For
> 
> ```
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
> ```
> 
> The algorithms don't support `in_place` sendbuf. The `MPIR_Ialltoallv_impl` function lists the exceptions to the algorithms. Same can be check in the other collectives/algos. I think the tests to be excluded are already listed here.

Thanks. That is helpful.

hzhou commented:
> 
> 
> For
> 
> ```
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_scattered.* -> xfail=ticket0
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_blocked.* -> xfail=ticket0
> * * * * *: ^alltoallv.* env=MPIR_CVAR_IALLTOALLV_INTRA_ALGORITHM=gentran_inplace.* -> xfail=ticket0
> ```

@tarudoodi These patterns are only intended for `alltoallv`, not `alltoallv0`, right?

hzhou commented:
@raffenet I have converted the script to python. It is ready for your review. Currently this has no impact until we update the jenkins tests to use the script inside mpich.

tarudoodi commented:
> @tarudoodi These patterns are only intended for `alltoallv`, not `alltoallv0`, right?

Correct, `alltoallv0` specifies a sendbuf.

raffenet commented:
I should mention I do not intend to backport this change to 3.4.x, so the Jenkins scripts should support both possibilities (but prefer the new Pythong script) for now.

hzhou commented:
> I should mention I do not intend to backport this change to 3.4.x, so the Jenkins scripts should support both possibilities (but prefer the new Pythong script) for now.

That should be fine. It is just 10 lines in a isolated branch.

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
Due to a bug misses following patterns:
```
testlist: test/mpi/errors/comm/testlist ...
        ^intercomm_abort .*     :ticket0
testlist: test/mpi/rma/testlist ...
        ^large_acc_flush_local .*       :issue3251
testlist: test/mpi/rma/testlist.dtp ...
        ^getfence1 [0-9]* arg=-type=.* arg=-count=16000000 .*   :ticket2965
        ^putfence1 [0-9]* arg=-type=.* arg=-count=16000000 .*   :ticket2965
        ^lock_contention_dt [0-9]* arg=-type=.* arg=-count=262144 .*    :ticket2965
        ^lock_dt.* [0-9]* arg=-type=.* arg=-count=262144 .*     :ticket2965
        ^lockall_dt.* [0-9]* arg=-type=.* arg=-count=262144 .*  :ticket2965
        ^epochtest [0-9]* arg=-type=.* arg=-count=262144 .*     :ticket2965
        ^lock_contention_dt [0-9]* arg=-type=.* arg=-count=32768 .*     :ticket2965
        ^lock_dt.* [0-9]* arg=-type=.* arg=-count=32768 .*      :ticket2965
        ^lockall_dt.* [0-9]* arg=-type=.* arg=-count=32768 .*   :ticket2965
        ^accfence1 [0-9]* arg=-type=.* arg=-count=262144 .*     :ticket2965
        ^accpscw1 [0-9]* arg=-type=.* arg=-count=262144 .*      :ticket2965
testlist: test/mpi/threads/comm/testlist ...
        ^idup_comm_gen .*       :ticket3794
        ^idup_nb .*     :ticket3794
        ^dup_leak_test .*iter=12345.*   :issue4595
testlist: test/mpi/threads/pt2pt/testlist ...
        ^ibsend .*      :ticket0
        ^mt_probe_sendrecv_huge .*      :issue4256
        ^mt_iprobe_sendrecv_huge .*     :issue4259
        ^mt_mprobe_sendrecv_huge .*     :issue4259
        ^mt_improbe_sendrecv_huge .*    :issue4259
```
Only following tests faild because of that:
```
not ok  - ./threads/comm/idup_comm_gen 4 Stack Trace | 3 min 0 sec
not ok  - ./threads/comm/idup_nb 4 Stack Trace | 3 min 0 sec 
```

, which means many of rma xfails have been fixed.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
@raffenet I updated the Jenkins scripts to use the in-repository `set_xfail.py` script when it exists, and found and fixed a few bugs (introduced during perl to python conversion). Now it seems working, could you re-approve the PR?

raffenet reviewed (commented):
I have to admit that after looking at the previous script, this new one seems much more complex to work with. Maybe it is because I use `sh` and `sed` much more in my day-to-day than regex, but I could see losing lots of time trying to debug or modify this script. I'd like others to chime in to get a consensus.

hzhou commented:
> I have to admit that after looking at the previous script, this new one seems much more complex to work with. Maybe it is because I use `sh` and `sed` much more in my day-to-day than regex, but I could see losing lots of time trying to debug or modify this script. I'd like others to chime in to get a consensus.

Fair enough. I just like to point out that the main motivation for rewriting the script is to make the debugging of `xfail.conf` easier rather than debugging this script.

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/ucx

hzhou commented:
test:mpich/ch4/most

hzhou commented:
@raffenet Are you still in the position of blocking this PR (https://github.com/pmodels/mpich/pull/4935#pullrequestreview-549639611)?

For your information, our Jenkins is already using this script (well, the Perl version of it) for a year now. I believe the concern of maintenance is overblown. For one, the xfail syntax is simple enough (simpler with the new script) it is not that often that we need to debug it. And for two, my experience is the opposite. I have lost many hours trying to debug the old shell scripts -- it was complicated since it generates secondary shell scripts to apply xfails and there was no easy way to debug individual line vs individual patterns nor there was any reasonable way to trace it. On the other hand, while the python script seems more complex than the shell scripts, it can be easily traced and debugged and directly read to understand.

raffenet commented:
> @raffenet Are you still in the position of blocking this PR ([#4935 (review)](https://github.com/pmodels/mpich/pull/4935#pullrequestreview-549639611))?
> 
> For your information, our Jenkins is already using this script (well, the Perl version of it) for a year now. I believe the concern of maintenance is overblown. For one, the xfail syntax is simple enough (simpler with the new script) it is not that often that we need to debug it. And for two, my experience is the opposite. I have lost many hours trying to debug the old shell scripts -- it was complicated since it generates secondary shell scripts to apply xfails and there was no easy way to debug individual line vs individual patterns nor there was any reasonable way to trace it. On the other hand, while the python script seems more complex than the shell scripts, it can be easily traced and debugged and directly read to understand.

No one seemed to care, so I guess we should just go ahead. I think the script is not very readable, but I'm probably not going to need to look at it at this point.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most
